Stricter backward compatibility check

raziel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45773 Stricter backward compatibility check**

Changes the function schema's backward compatibility check to be stricter to comply with C++ API backwards compatibility capabilities.

See #45784 for description

Differential Revision: [D24089751](https://our.internmc.facebook.com/intern/diff/D24089751/)

Commit:
Stricter backward compatibility check

Changes the function schema's backward compatibility check to be stricter to comply with C++ API backwards compatibility capabilities.

Differential Revision: [D24089751](https://our.internmc.facebook.com/intern/diff/D24089751/)

[ghstack-poisoned]

Commit:
Update on "Stricter backward compatibility check"


Changes the function schema's backward compatibility check to be stricter to comply with C++ API backwards compatibility capabilities.

Differential Revision: [D24089751](https://our.internmc.facebook.com/intern/diff/D24089751/)

[ghstack-poisoned]

mruberry commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -156,18 +156,22 @@ struct FunctionSchema {
-  //      this schema must provide default values.
+  // Checks whether this schema is backward compatible with the old one.
+  // The following conditions must be true:
+  // * The function's structure of old and new match (e.g. name,
```
I think this could be a little clearer. A complete enumeration would be something like this, right?

- [Output Narrowing] The new schema's output type must be the same class or inherit from the old schema's output type
- [Arg Count] The new schema must have at least as many arguments as the old schema
- [Default Values] Every new argument must have a default value
- [ Arg Compatibility] Every argument in the old schema has a corresponding argument in the new schema that..
  - is at the same position
  - has the same name
  - has the same type, or the old argument's type inherits from the new argument's type

In the current description I think it's unclear what "matching" means.

raziel replied:
Thanks for the comment. I think what you suggest makes sense.

One thing though: I think simplifying the explanation just in terms of arguments as the list of all positional and kwargs makes it a 
bit simpler to think and avoid dealing with the individual concepts.
E.g. "The new schema must have exactly as many positional arguments as the old schema" is not quite the case since since a kwargs in old may have moved to be a positional in new (i.e. in the case f_new(a, b, *, c) => f_old(a, *, b, c) , f_new would have 2 positional vs f_old's 1). That's why there's not a direct requirement on number of positional arguments being the same.

mruberry commented on test/test_function_schema.py:
```diff
@@ -27,11 +27,14 @@ def test_backward_compatible_args(self):
         self.assertFalse(old_schema.is_backward_compatible_with(new_schema))
 
     def test_backward_compatible_kwargs(self):
-        old_schema = parse_schema('any(Tensor self, *, Tensor out) -> Tensor')
```
The current test organization is difficult to read and maintain.

I would expect that a single schema is created and each of the properties are tested sequentially with a comment starting the block that tests them. For example:

```
# Tests output narrowing
  <new schema where output is wider fails>
  <new schema where output is narrow passes>

# Tests arg count
  < new schema with fewer args fails>
```

raziel replied:
Agreed. Updated tests with new organization and comments.

mruberry commented:
This makes a lot of sense, @raziel, and overall I think this PR looks pretty good. I made a couple suggestions to clarify a comment and update the existing tests for readability.

Commit:
Update on "Stricter backward compatibility check"


Changes the function schema's backward compatibility check to be stricter to comply with C++ API backwards compatibility capabilities.

See #45784 for description

Differential Revision: [D24089751](https://our.internmc.facebook.com/intern/diff/D24089751/)

[ghstack-poisoned]

raziel commented:
Thanks @mruberry for the review and comments.
Overall I tried to keep consistency with the original code and comments, but I agree with your observations.
PR updated to address them.

mruberry commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -156,18 +156,27 @@ struct FunctionSchema {
+  //      or inherit from the old schema's output type
+  // [Argument count] The new schema must have at least as many arguments as
+  //      the old schema (considering the list of positional and kwargs).
+  // [Arg Compatibility] Every argument in the old schema has a corresponding
```
I like your organization. Should "Arg Compatibility" also cover the movement of args from positional to keyword-only? Here's the cases I'm thinking of

```
foo(a, b, *, c) -> foo(a, b, c)
# this seems OK?

foo(a, b, *, c) -> foo(a, *, b, c)
# this seems not OK
```

raziel replied:
Yes, you're right the second case "kwarg to positional being BC" is covered in the tests:
# BC: A new schema where old kwargs becomes positional: foo_new(a, b, *, c) -> foo_old(a, *, b, c)
Along with:
# BC: A new schema where all old kwargs become positional: foo_new(a, b, c) -> foo_old(a, *, b, c)

And indeed the opposite ("positional to kwarg not being BC") is not covered neither in the docs nor tests.

I just added a couples of test to cover what you mention:
# No-BC: A new schema where an old positional argument becomes kwarg: foo_new(a, *, b, c) -> foo_old(a, b, *, c)
# No-BC: A new schema where many old positional arguments become kwarg: foo_new(a, *, b, c) -> foo_old(a, b, c)

Also updated the docs to reflect this asymmetry.

Thanks!

Commit:
Update on "Stricter backward compatibility check"


Changes the function schema's backward compatibility check to be stricter to comply with C++ API backwards compatibility capabilities.

See #45784 for description

Differential Revision: [D24089751](https://our.internmc.facebook.com/intern/diff/D24089751/)

[ghstack-poisoned]

mruberry reviewed (approved):
Cool update, @raziel! And thank you for updating those old tests.
